server/app/assets/javascripts/admin_predicate_configuration.ts

```diff
@@ -0,0 +1,479 @@
+
+    const valueRow = (event.target as HTMLElement).closest(
+      '.predicate-config-value-row',
+    )
+    valueRow.remove()
+  }
+
+  /**
+   * Filter the operators available for each scalar type based on the current scalar selected.
+   *   @param {HTMLSelectElement} scalarDropdown The element to filter the operators for.
+   *   @param {string} selectedScalarType The tyoe of the selected option
+   *   @param {string} selectedScalarValue The value of the selected option
+   */
+  filterOperators(
+    scalarDropdown: HTMLSelectElement,
+    selectedScalarType: string,
+    selectedScalarValue: string,
+  ) {
+    const questionId = this.getQuestionId(scalarDropdown)
+    const operatorDropdown = this.getElementWithQuestionId(
+      '.cf-operator-select',
+      questionId,
+      'select',
+    ) as HTMLSelectElement
+
+    Array.from(operatorDropdown.options).forEach((operatorOption) => {
+      operatorOption.classList.toggle(
+        'hidden',
+        !this.shouldHideOperator(
+          selectedScalarType,
+          selectedScalarValue,
+          operatorOption,
+        ),
+      )
+    })
+  }
+
+  /**
+   * Logic that decides if a operator should be hidden.
+   *   @param {string} selectedScalarType The tyoe of the selected option
+   *   @param {string} selectedScalarValue The value of the selected option
+   *   @param {HTMLOptionElement} operatorOption The operator to check if we should hide.
+   * @return {boolean} If the operator should be hidden
+   */
+  private shouldHideOperator(
+    selectedScalarType: string,
+    selectedScalarValue: string,
+    operatorOption: HTMLOptionElement,
+  ): boolean {
+    // If this operator is not for the currently selected type, hide it.
+    return (
+      // Special case for SELECTION scalars (which are of type STRING):
+      // do not include EQUAL_TO or NOT_EQUAL_TO. This is because we use a set of checkbox
+      // inputs for values for multi-option question predicates, which works well for list
+      // operators such as ANY_OF and NONE_OF. Because you can achieve the same functionality
+      // of EQUAL_TO with ANY_OF and NOT_EQUAL_TO with NONE_OF, we made a technical choice to
+      // exclude these operators from single-select predicates to simplify the code on both
+      // the form processing side and on the admin user side.
+      !(selectedScalarType in operatorOption.dataset) ||
+      (selectedScalarValue.toUpperCase() === 'SELECTION' &&
+        (operatorOption.value === 'EQUAL_TO' ||
+          operatorOption.value === 'NOT_EQUAL_TO'))
+    )
+  }
+
+  private getQuestionId(element: HTMLSelectElement) {
+    return (element.closest(`[data-question-id]`) as HTMLElement)?.dataset
+      .questionId
+  }
+
+  private getElementWithQuestionId(
+    selector: string,
+    questionId: string,
+    childSelector: string,
+  ) {
+    return document.querySelector(
+      `${selector}[data-question-id="${questionId}"] ${childSelector}`,
+    )
+  }
+}
+
```

Comment:
note: all code below this comment is moved from `main.ts` and otherwise unchanged

Reply:
> How should I interpret this duplication/subtle-differentiation? 

As an intermediate step toward removing the old code with top-level functions.

> Can the "new" ones be named (even temp) in a way that makes sense for readability and to reduce errors?

What changes do naming would be helfpul?

Reply:
moved the legacy code into a separate file